Sources/SwiftGRPCNIOInteroperabilityTests/InteroperabilityTestCases.swift

```diff
@@ -0,0 +1,679 @@
+    try assertEqual(call.status.map { $0.code }, .unimplemented)
+  }
+}
+
+/// This test verifies that a request can be cancelled after metadata has been sent but before
+/// payloads are sent.
+///
+/// Server features:
+/// - StreamingInputCall
+///
+/// Procedure:
+/// 1. Client starts StreamingInputCall
+/// 2. Client immediately cancels request
+///
+/// Client asserts:
+/// - Call completed with status CANCELLED
+class CancelAfterBegin: InteroperabilityTest {
+  func run(using connection: GRPCClientConnection) throws {
+    let client = Grpc_Testing_TestServiceService_NIOClient(connection: connection)
+    let call = client.streamingInputCall()
+    call.cancel()
+
+    try assertEqual(call.status.map { $0.code }, .cancelled)
+  }
+}
+
+/// This test verifies that a request can be cancelled after receiving a message from the server.
+///
+/// Server features:
+/// - FullDuplexCall
+///
+/// Procedure:
+/// 1. Client starts FullDuplexCall with
+///    ```
+///    {
+///        response_parameters:{
+///            size: 31415
+///        }
+///        payload:{
+///            body: 27182 bytes of zeros
+///        }
+///    }
+///    ```
+/// 2. After receiving a response, client cancels request
+///
+/// Client asserts:
+/// - Call completed with status CANCELLED
+class CancelAfterFirstResponse: InteroperabilityTest {
+  func run(using connection: GRPCClientConnection) throws {
+    let client = Grpc_Testing_TestServiceService_NIOClient(connection: connection)
+
+    let promise = client.connection.channel.eventLoop.makePromise(of: Void.self)
+
+    let call = client.fullDuplexCall { _ in
+      promise.succeed(())
+    }
+
+    promise.futureResult.whenSuccess {
+      call.cancel()
+    }
+
+    let request = Grpc_Testing_StreamingOutputCallRequest.with { request in
+      request.responseParameters = [.size(31_415)]
+      request.payload = .zeros(count: 27_182)
+    }
+
+    call.sendMessage(request, promise: nil)
+
+    try assertEqual(call.status.map { $0.code }, .cancelled)
+  }
+}
+
+/// This test verifies that an RPC request whose lifetime exceeds its configured timeout value
+/// will end with the DeadlineExceeded status.
+///
+/// Server features:
+/// - FullDuplexCall
+///
+/// Procedure:
+/// 1. Client calls FullDuplexCall with the following request and sets its timeout to 1ms
+///    ```
+///    {
+///        payload:{
+///            body: 27182 bytes of zeros
+///        }
+///    }
+///    ```
+/// 2. Client waits
+///
+/// Client asserts:
+/// - Call completed with status DEADLINE_EXCEEDED.
+class TimeoutOnSleepingServer: InteroperabilityTest {
+  func run(using connection: GRPCClientConnection) throws {
+    let client = Grpc_Testing_TestServiceService_NIOClient(connection: connection)
+
+    let callOptions = CallOptions(timeout: try .milliseconds(1))
+    let call = client.fullDuplexCall(callOptions: callOptions) { _ in }
+
+    try assertEqual(call.status.map { $0.code }, .deadlineExceeded)
```

Comment:
Nice, I had completely forgotten that we actually enforce deadlines. Do you remember if we enforce them in the client, on the server, or both?

Reply:
Only on the client at the moment. We do send the timeout to the server though, so other implementations which support it can use it.